engine/classes/Elgg/Notifications/NotificationEventHandler.php

```diff
@@ -0,0 +1,356 @@
+
+/**
+ * Notification Event Handler handles preparation of a notification
+ *
+ * @since 4.0
+ */
+class NotificationEventHandler {
+
+	/** @var NotificationEvent */
+	protected $event;
+
+	/** @var NotificationsService */
+	protected $service;
+
+	/** @var array */
+	protected $subscriptions = [];
+
+	/** @var array */
+	protected $params = [];
+
+	/**
+	 * Constructor
+	 *
+	 * @param NotificationEvent    $event   event to handle
+	 * @param NotificationsService $service service that handles the events
+	 * @param array                $params  additional params for event handling
+	 */
+	public function __construct(
+			NotificationEvent $event,
+			NotificationsService $service,
+			array $params = []
+	) {
+		$this->event = $event;
+		$this->service = $service;
+		$this->params = $params;
+	}
+	
+	/**
+	 * Process the event
+	 *
+	 * @return array delivery information
+	 */
+	final public function send(): array {
+		$deliveries = [];
+		
+		$params = $this->params;
+		$params['handler'] = $this;
+		$params['event'] = $this->event;
+		$params['subscriptions'] = $this->prepareSubscriptions();
+
+		// return false to stop the default notification sender
+		if (_elgg_services()->hooks->trigger('send:before', 'notifications', $params, true)) {
+			$deliveries = $this->sendNotifications($params['subscriptions'], $params);
+		}
+		
+		$params['deliveries'] = $deliveries;
+		_elgg_services()->hooks->trigger('send:after', 'notifications', $params);
+		
+		return $deliveries;
+	}
+	
+	/**
+	 * Returns subscriptions
+	 *
+	 * @return array
+	 */
+	final protected function prepareSubscriptions(): array {
+		$subscriptions = $this->getSubscriptions();
+		
+		$params = [
+			'event' => $this->event,
+			'methods' => $this->getMethods(),
+			'methods_override' => (array) elgg_extract('methods_override', $this->params, []),
+		];
+		$subscriptions = _elgg_services()->hooks->trigger('get', 'subscriptions', $params, $subscriptions);
+		
+		return _elgg_services()->subscriptions->filterSubscriptions($subscriptions, $this->event);
+	}
+	
+	/**
+	 * Returns subscriptions for the event
+	 *
+	 * @return array
+	 */
+	public function getSubscriptions(): array {
+		return _elgg_services()->subscriptions->getNotificationEventSubscriptions($this->event, $this->getMethods());
+	}
+	
+	/**
+	 * Returns methods to be used for this notification
+	 *
+	 * @return array
+	 */
+	public function getMethods(): array {
+		return $this->service->getMethods();
+	}
+	
+	/**
+	 * Sends the notifications based on subscriptions
+	 *
+	 * Returns an array in the form:
+	 * <code>
+	 * [
+	 *    25 => [
+	 *      'email' => true,
+	 *      'sms' => false,
+	 *    ],
+	 *    55 => [],
+	 * ]
+	 * </code>
+	 *
+	 * @param array $subscriptions Subscriptions for this event
+	 * @param array $params        Default notification parameters
+	 *
+	 * @return array
+	 */
+	protected function sendNotifications($subscriptions, array $params = []) {
+
+		$methods = $this->service->getMethods();
```

Comment:
- why not use `$this->getMethods()` as you do above in `getSubscriptions()`
- why store in a variable if it's only used in the line below
- if the variable is needed, maybe rename it to avoid confusion with line 130

Reply:
changed